RAT/utils/plotting.py

```diff
@@ -0,0 +1,190 @@
+import numpy as np
+from RAT.rat_core import PlotEventData, makeSLDProfileXY
+
+
+class Figure:
+    """
+    Creates a plotting figure.
+    """
+
+    def __init__(self, row: int = 1, col: int = 2):
+        """
+        Initializes the figure and the subplots.
+
+        Parameters
+        ----------
+        row : int
+              The number of rows in subplot
+        col : int
+              The number of columns in subplot
+        """
+        self._fig, self._ax = \
+            plt.subplots(row, col, num="Reflectivity Algorithms Toolbox (RAT)")
+        plt.show(block=False)
+        self._esc_pressed = False
+        self._close_clicked = False
+        self._fig.canvas.mpl_connect("key_press_event",
+                                     self._process_button_press)
+        self._fig.canvas.mpl_connect('close_event',
+                                     self._close)
+
+    def wait_for_close(self):
+        """
+        Waits for the user to close the figure
+        using the esc key.
+        """
+        while not (self._esc_pressed or self._close_clicked):
+            plt.waitforbuttonpress(timeout=0.005)
+        plt.close(self._fig)
+
+    def _process_button_press(self, event):
+        """
+        Process the key_press_event.
+        """
+        if event.key == 'escape':
+            self._esc_pressed = True
+
+    def _close(self, _):
+        """
+        Process the close_event.
+        """
+        self._close_clicked = True
+
+
+def plot_errorbars(ax, x, y, err, onesided, color):
+    """
+    Plots the error bars.
+
+    Parameters
+    ----------
+    ax : matplotlib.axes._axes.Axes
+            The axis on which to draw errorbars
+    x : np.ndarray
+        The shifted data x axis data
+    y : np.ndarray
+        The shifted data y axis data
+    err : np.ndarray
+            The shifted data e data
+    onesided : bool
+        A boolean to indicate whether to draw one sided errorbars
+    color : str
+            The hex representing the color of the errorbars
+    """
+    y_error = [[0]*len(err), err] if onesided else err
+    ax.errorbar(x=x,
+                y=y,
+                yerr=y_error,
+                fmt='none',
+                ecolor=color,
+                elinewidth=1,
+                capsize=0)
+    ax.scatter(x=x, y=y, s=3, marker="o", color=color)
+
+
+def plot_ref_sld(data: PlotEventData, fig: Figure = None, delay: bool = True):
+    """
+    Clears the previous plots and updates the ref and SLD plots.
+
+    Parameters
+    ----------
+    data : PlotEventData
+           The plot event data that contains all the information
+           to generate the ref and sld plots
+    fig : Figure
+          The figure class that has two subplots
+    delay : bool
+            Controls whether to delay 0.005s after plot is created
+
+    Returns
+    -------
+    fig : Figure
+          The figure class that has two subplots
+    """
+    if fig is None:
+        fig = Figure()
```

Comment:
When a figure is provided, please check if the figure has the correct number of axes i.e. 2 and if not, recreate the subplot with the correct subplots `fig.fig.subplot(1, 2)`

Reply:
I pushed the changes. But it requires the figure to be closed first and then recreated.